Document the caching of autotracked properties (#1474)
ember-learn/guides-source: guides/release/in-depth-topics/autotracking-in-depth.md

}
```

## Caching of tracked properties

> In contrast with computed properties from pre-Octane, autotracked properties
> are not cached. This means that an autotracked property will always be
> recomputed, regardless of whether its dependencies have changed. The following
> example shows how this works:

Reviewer: I think it's better that we don't use the term `autotracked property`, in case this may cause a reader to think it's different from `tracked property`.
```suggestion
In contrast to computed properties from pre-Octane, tracked properties are
not cached. A tracked property can also be recomputed even though its
dependencies haven't changed. The following example shows this behavior:
```

Author: Makes sense.

```js
import { tracked } from '@glimmer/tracking';

let count = 0;

> class Person {
>   @tracked firstName = 'Jen';
>   @tracked lastName = 'Weber';
> 
>   get fullName() {
>     ++count;
>     return `${this.firstName} ${this.lastName}`;
>   }
> }
> 
> let person = new Person();
> 
> console.log(person.fullName); // Jen Weber
> console.log(count); // 1;
> console.log(person.fullName); // Jen Weber
> console.log(count); // 2;
> 
> person.firstName = 'Jennifer';
> 
> console.log(person.fullName); // Jennifer Weber
> console.log(count); // 3;

Reviewer: To help address https://github.com/ember-learn/guides-source/issues/1480, let's not use a `Person` class. Can we use the following example instead?

```suggestion
class Photo {
  @tracked width = 600;
  @tracked height = 400;

  get aspectRatio() {
    count++;
    return this.width / this.height;
  }
}

let photo = new Photo();

console.log(photo.aspectRatio); // 1.5
console.log(count); // 1;
console.log(photo.aspectRatio); // 1.5
console.log(count); // 2;

photo.width = 800;

console.log(photo.aspectRatio); // 2
console.log(count); // 3;
```

Author: Cool, agree.

```

> This works fine in most cases. But sometimes you want a bit more control over
> the recomputation, for instance if the computation that happens in the getter
> is very expensive. In that case, you want to cache the value in between calls
> when the dependencies haven't changed. You want to recompute only if one of the
> dependencies has been modified.

Reviewer: ```suggestion
Recomputing is fine in most cases. If the computation that happens in the getter
is very expensive, however, you will want to cache the value and retrieve it when
the dependencies haven't changed. You want to recompute only if one of the
dependencies has been modified.
```

Author: Good.

> The way to achieve this is through Ember's [cache
> API](https://github.com/emberjs/rfcs/blob/master/text/0615-autotracking-memoization.md).
> Using `createCache`, we can cache an autotracked property. The property can
> then be consumed with the `getValue` function. The way it works is as follows:

Reviewer: I added a link to the polyfill.

According to the [RFC](https://github.com/emberjs/rfcs/blob/master/text/0615-autotracking-memoization.md#usage), I think `createCache` receives a function. I updated your Guides text to make this clear:

```suggestion
Ember's [cache API](https://github.com/ember-polyfills/ember-cache-primitive-polyfill) lets you cache a getter in 2 steps.
First, you pass a function that is costly to compute to `createCache`.
Then, in the getter, you call the function with `getValue` and return its value.

With these 2 steps in mind, let's introduce caching to `aspectRatio`:
```

Author: Nice!

```js
import { tracked } from '@glimmer/tracking';
import { createCache, getValue } from '@glimmer/tracking/primitives/cache';

let count = 0;

> class Person {
>   @tracked firstName = 'Jen';
>   @tracked lastName = 'Weber';
> 
>   // `#` indicates a private field on the class.
>   #fullNameCache = createCache(() => {
>     ++count;
>     return `${this.firstName} ${this.lastName}`;
>   })
> 
>   get fullName() {
>     return getValue(this.#fullNameCache);
>   }
> }
> 
> let person = new Person();
> 
> console.log(person.fullName); // Jen Weber
> console.log(count); // 1;
> console.log(person.fullName); // Jen Weber
> console.log(count); // 1;
> 
> person.firstName = 'Jennifer';
> 
> console.log(person.fullName); // Jennifer Weber
> console.log(count); // 2;

Reviewer: ```suggestion
class Photo {
  @tracked width = 600;
  @tracked height = 400;

  // `#` indicates a private field on the class
  #aspectRatioCache = createCache(() => {
    count++;
    return this.width / this.height;
  })

  get aspectRatio() {
    return getValue(this.#aspectRatioCache);
  }
}

let photo = new Photo();

console.log(photo.aspectRatio); // 1.5
console.log(count); // 1;
console.log(photo.aspectRatio); // 1.5
console.log(count); // 1;

photo.width = 800;

console.log(photo.aspectRatio); // 2
console.log(count); // 2;
```

```

> 

Reviewer: ```suggestion

From the value of `count`, we see that, this time, `aspectRatio` was calculated only twice.
```
